File: storage-common/src/main/java/io/github/solusmods/eternalcore/storage/impl/network/s2c/SyncWorldStoragePayload.java
package io.github.solusmods.eternalcore.storage.impl.network.s2c;


import dev.architectury.networking.NetworkManager;
import dev.architectury.utils.Env;
import net.minecraft.nbt.CompoundTag;
import net.minecraft.network.FriendlyByteBuf;
import net.minecraft.network.codec.StreamCodec;
import net.minecraft.network.protocol.common.custom.CustomPacketPayload;
import net.minecraft.resources.ResourceLocation;

import static io.github.solusmods.eternalcore.network.ModuleConstants.MOD_ID;

public record SyncWorldStoragePayload(
        boolean isUpdate,
        CompoundTag storageTag
) implements StorageSyncPayload {
    public static final Type<SyncWorldStoragePayload> TYPE = new Type<>(ResourceLocation.fromNamespaceAndPath(MOD_ID, "sync_world_storage"));
    public static final StreamCodec<FriendlyByteBuf, SyncWorldStoragePayload> STREAM_CODEC = CustomPacketPayload.codec(SyncWorldStoragePayload::encode, SyncWorldStoragePayload::new);

    public SyncWorldStoragePayload(FriendlyByteBuf buf) {
        this(buf.readBoolean(), buf.readNbt());
    }

    public void encode(FriendlyByteBuf buf) {
        buf.writeBoolean(isUpdate);
        buf.writeNbt(storageTag);
    }

    public void handle(NetworkManager.PacketContext context) {
        if (context.getEnvironment() != Env.CLIENT) return;
        context.queue(() -> ClientAccess.handle(this));
    }

    @Override
    public Type<SyncWorldStoragePayload> type() {
        return TYPE;
    }
}

File: storage-common/src/main/kotlin/io/github/solusmods/eternalcore/storage/impl/network/s2c/SyncWorldStoragePayload.kt
package io.github.solusmods.eternalcore.storage.impl.network.s2c

import dev.architectury.networking.NetworkManager
import dev.architectury.utils.Env
import io.github.solusmods.eternalcore.network.ModuleConstants
import net.minecraft.nbt.CompoundTag
import net.minecraft.network.FriendlyByteBuf
import net.minecraft.network.codec.StreamCodec
import net.minecraft.network.protocol.common.custom.CustomPacketPayload
import net.minecraft.resources.ResourceLocation


@JvmRecord
data class SyncWorldStoragePayload(
    override val isUpdate: Boolean,
    override val storageTag: CompoundTag
) : StorageSyncPayload {
    constructor(buf: FriendlyByteBuf) : this(buf.readBoolean(), buf.readNbt()!!)

    fun encode(buf: FriendlyByteBuf) {
        buf.writeBoolean(isUpdate)
        buf.writeNbt(storageTag)
    }

    fun handle(context: NetworkManager.PacketContext) {
        if (context.environment != Env.CLIENT) return
        context.queue { ClientAccess.handle(this) }
    }

    override fun type(): CustomPacketPayload.Type<SyncWorldStoragePayload> {
        return TYPE
    }

    companion object {
        @JvmField
        val TYPE: CustomPacketPayload.Type<SyncWorldStoragePayload> =
            CustomPacketPayload.Type<SyncWorldStoragePayload>(
                ResourceLocation.fromNamespaceAndPath(ModuleConstants.MOD_ID, "sync_world_storage")
            )
        @JvmField
        val STREAM_CODEC: StreamCodec<FriendlyByteBuf, SyncWorldStoragePayload> =
            CustomPacketPayload.codec<FriendlyByteBuf, SyncWorldStoragePayload>(
                { obj: SyncWorldStoragePayload, buf: FriendlyByteBuf -> obj.encode(buf) },
                { buf: FriendlyByteBuf ->
                    SyncWorldStoragePayload(
                        buf
                    )
                })
    }
}